### PR TITLE
Madeline Bennett - Pheonix Class

### PR DESCRIPTION
**Pull request by MBsea21**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,45 @@
...
+        <button className="like"onClick={onLikeClicked}>{likeButtonText}</button>
       </section>
     </div>
   );
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender:PropTypes.string.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:35]: Nit: key colons should be followed by a space, throughout.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,45 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp'
+
+const ChatEntry = ({id,sender,body,timeStamp,liked, chatLikeClicked, uniqueParticipants}) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: Nit: commas should be followed by a space.

Comment (anselrognlie) [reply src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,45 @@
...
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={participantClass}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time"><TimeStamp time= {timeStamp}></TimeStamp></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:26]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Note that the `TimeStamp` component doesn't receive children components (it resembles "void" HTML tags such as `input` or `img` rather than tags like `p` or `a` that can have children). For such components, prefer to write them using "self-closing" style (note the `/>` to complete the tag rather than a regular `>`).
```html
<TimeStamp time={timeStamp} />
```

Nit: don't leave spaces around attribute `=`

File: src/components/ChatEntry.jsx [hunk 4]
```diff
@@ -1,20 +1,45 @@
...
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender:PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp:PropTypes.string.isRequired,
+  liked:PropTypes.bool.isRequired,
+  chatLikeClicked:PropTypes.func.isRequired,
+  uniqueParticipants:PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:40]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did.

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests (specifically for `chatLikeClicked`), be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests (like you did for `uniqueParticipants`) to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender:PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp:PropTypes.string.isRequired,
+      liked:PropTypes.bool.isRequired,
+    })
+  ).isRequired,
+  chatLikeClicked:PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:37]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+          uniqueParticipants={uniqueParticipants}
+        />
+      );
+    });
+  };
+  
+  return ( <ul className="chat-log"> {getIndividualChats(entries)} </ul>);
+};
+
+ChatLog.propTypes ={
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:27]: Nit: leave space around assignment `=`

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import ChatEntry from './ChatEntry'
+
+
+const ChatLog = ({entries, chatLikeClicked, uniqueParticipants}) => {
+  const getIndividualChats = (entries) => {
+    return entries.map((chatEntry) => {
+      return (
+        <ChatEntry
+          key={chatEntry.id}
+          id={chatEntry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:12]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+import ChatEntry from './ChatEntry'
+
+
+const ChatLog = ({entries, chatLikeClicked, uniqueParticipants}) => {
+  const getIndividualChats = (entries) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:7]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping using a helper function as you did here, we could assign the map result directly to a local variable, or the map expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component. Calling the helper function in the `return` or using the expression directly can make debugging a little trickier, since more stuff happens in the single `return` expression, but any of those approaches will work fine.

File: src/components/ChatLog.jsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+          timeStamp={chatEntry.timeStamp}
+          liked={chatEntry.liked}
+          chatLikeClicked={chatLikeClicked}
+          uniqueParticipants={uniqueParticipants}
+        />
+      );
+    });
+  };
+  
+  return ( <ul className="chat-log"> {getIndividualChats(entries)} </ul>);
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:24]: Since you're using an `ul` tag to list your chat entries, the child of this should be `li` tags. This can be accomplished either by having `ChatEntry` use `li` as its top-level tag (yours uses `div`) or by wrapping each `ChatEntry` in an `li` in the mapping step. We should aim to generate valid HTML once all the JSX components have been expanded to actual tags.

Comment (anselrognlie) [reply src/components/ChatLog.jsx:24]: Nit: this `return` easily fits on one line, so no need to wrap the expression in `()`.

File: src/App.jsx [hunk 10]
```diff
@@ -1,12 +1,45 @@
...
+import ChatLog from './components/ChatLog.jsx'
+import chatsJson from './data/messages.json';
+import { useState } from 'react';
+import AppTitle from './components/AppTitle.jsx';
+
+
 
 const App = () => {
+  const [chatsData, setChatData] = useState(chatsJson);
+  const [likeCount,setLikeCount] = useState(0);
```
Comment (anselrognlie) [inline src/App.jsx:12]: 👀 It's not necessary to introduce additional state to track the total number of likes, and really we would prefer to avoid doing so. The total number of likes can be calculated from the message data, since each message tracks whether it's liked. The total number of likes is an aggregate property of the message data.

It can be tempting to introduce additional state since it can seem inefficient to iterate over all the message data just to calculate that value. But keep in mind that we have to do this anyway just to render the message data into components, so the additional cost is minimal, and avoids potential problems that could lead to states getting out of sync. For a small application like this, the risk is small, but as apps get more complex, we should look for ways to reduce the state used as much as possible.

File: src/App.jsx [hunk 11]
```diff
@@ -1,12 +1,45 @@
...
+
+
+  const handleLikeClicked = (id) => {
+    setChatData((chatsData)=> chatsData.map((chatEntry) => {
+      if (chatEntry.id === id) {
+        if (chatEntry.liked === true) {
+          setLikeCount(likeCount - 1);}
+        else {
+          setLikeCount(likeCount + 1);
+        }
```
Comment (anselrognlie) [inline src/App.jsx:23]: If we exclude the additional state to track the total likes, we wouldn't need this logic here. We would need a similar check in whatever logic we use to calculate the total on the next render, but being able to remove the extra state makes moving the logic worth it.

File: src/App.jsx [hunk 12]
```diff
@@ -1,12 +1,45 @@
...
+
+
 
 const App = () => {
+  const [chatsData, setChatData] = useState(chatsJson);
+  const [likeCount,setLikeCount] = useState(0);
+  const uniqueParticipants = [...new Set(chatsData.map((chat)=>chat.sender))];
+
+
+  const handleLikeClicked = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:16]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 13]
```diff
@@ -1,12 +1,45 @@
...
+    setChatData((chatsData)=> chatsData.map((chatEntry) => {
```
Comment (anselrognlie) [inline src/App.jsx:17]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

Nit: consistent spacing around `=>`
```js
    setChatData((chatsData) => chatsData.map((chatEntry) => {
```

Comment (anselrognlie) [reply src/App.jsx:17]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,45 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp'
+
+const ChatEntry = ({id,sender,body,timeStamp,liked, chatLikeClicked, uniqueParticipants}) => {
+  const onLikeClicked = () => {
+    chatLikeClicked(id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 15]
```diff
@@ -1,20 +1,45 @@
...
+    if (sender === uniqueParticipants[0]){
+      return 'chat-entry local'
+    } else {
+      return 'chat-entry remote'
+    }
+  };
+
+  const participantClass = getParticipantClass();
+
+  const likeButtonText = !liked ? '🤍' : '❤️';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:19]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 16]
```diff
@@ -1,20 +1,45 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={participantClass}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time"><TimeStamp time= {timeStamp}></TimeStamp></p>
+        <button className="like"onClick={onLikeClicked}>{likeButtonText}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:27]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

Nit: spacing between attributes
```html
        <button className="like" onClick={onLikeClicked}>{likeButtonText}</button>
```

File: src/App.jsx [hunk 17]
```diff
@@ -1,12 +1,45 @@
...
+import chatsJson from './data/messages.json';
+import { useState } from 'react';
+import AppTitle from './components/AppTitle.jsx';
+
+
 
 const App = () => {
+  const [chatsData, setChatData] = useState(chatsJson);
+  const [likeCount,setLikeCount] = useState(0);
+  const uniqueParticipants = [...new Set(chatsData.map((chat)=>chat.sender))];
```
Comment (anselrognlie) [inline src/App.jsx:13]: Nice use of `map` to extract the sender information, then uniquing that with a `Set` to get the participants in the conversation. We can confirm with the docs that iterating over the `Set` contents will occur in insertion order, meaning the resulting list of senders, however many there are, will be in the order they participated in the conversation.

File: src/components/ChatEntry.jsx [hunk 18]
```diff
@@ -1,20 +1,45 @@
...
+const ChatEntry = ({id,sender,body,timeStamp,liked, chatLikeClicked, uniqueParticipants}) => {
+  const onLikeClicked = () => {
+    chatLikeClicked(id);
+  };
+  const getParticipantClass = () => {
+    if (sender === uniqueParticipants[0]){
+      return 'chat-entry local'
+    } else {
+      return 'chat-entry remote'
+    }
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:14]: 👍 We didn't receive any explicit data indicating which of the senders is local and which is remote, so deciding to treat the first sender as local is a reasonable decision.

File: src/components/AppTitle.jsx [hunk 19]
```diff
@@ -0,0 +1,13 @@
+// import { useState } from 'react';
+import PropTypes from 'prop-types';
+
+const AppTitle = ({uniqueParticipants}) => {
+  return (
+    <h1>Chat {uniqueParticipants.length > 0 ? `Between ${uniqueParticipants[0]} and ${uniqueParticipants[1]}` : ``}</h1>
```
Comment (anselrognlie) [inline src/components/AppTitle.jsx:6]: Nice way to build the header, only naming the senders when there's enough information to do so without error. In a full application, we probably would have received more explicit information about the participants in the conversation, but given the data limitations, this is a reasonable compromise. We could even add additional logic to handle more than two participants!

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
